npm/pkg/dataplane/policies/chain-management_linux.go

```diff
@@ -294,6 +243,235 @@ func (pMgr *PolicyManager) bootup(_ []string) error {
+		klog.Info(msg)
+		metrics.SendLog(util.IptmID, msg, metrics.DonotPrint)
+		util.SetIptablesToNft()
+		return nil
+	}
+
+	if err != nil {
+		msg := "failed to detect iptables version on first attempt. error running iptables-nft-save. will try detecting using iptables-legacy-save. err: %s"
+		metrics.SendErrorLogAndMetric(util.IptmID, msg, err.Error())
+	}
+
+	klog.Info("second attempt detecting iptables version. running: iptables-legacy-save -t mangle")
+	lCmd := pMgr.ioShim.Exec.Command(util.IptablesSaveLegacy, "-t", "mangle")
+	loutput, err := lCmd.CombinedOutput()
+	if err == nil && strings.Contains(string(loutput), "KUBE-IPTABLES-HINT") || strings.Contains(string(loutput), "KUBE-KUBELET-CANARY") {
+		msg := "detected iptables version on second attempt. found KUBE chains in legacy tables. NPM will use iptables-legacy"
+		klog.Info(msg)
+		metrics.SendLog(util.IptmID, msg, metrics.DonotPrint)
+		util.SetIptablesToLegacy()
+		return nil
+	}
+
+	if err != nil {
+		msg := "failed to detect iptables version on second attempt. error running iptables-legacy-save. will try detecting using kernel version. err: %s"
+		metrics.SendErrorLogAndMetric(util.IptmID, msg, err.Error())
+	}
+
+	klog.Info("third attempt detecting iptables version. getting kernel version")
+	kernelRelease := ""
+	if pMgr.debug {
+		// for testing purposes
+		kernelRelease = pMgr.debugKernelVersion
+	} else {
+		kernelRelease = util.KernelRelease()
+	}
+	kernelVersion := strings.Split(kernelRelease, ".")[0]
+	if kernelVersion == "" {
+		metrics.SendErrorLogAndMetric(util.IptmID, "failed to detect iptables version on third attempt. error getting kernel version")
+		return errDetectingIptablesVersion
+	}
+
+	majorVersion, err := strconv.Atoi(kernelVersion)
+	if err != nil {
+		metrics.SendErrorLogAndMetric(util.IptmID, "failed to detect iptables version on third attempt. error converting kernel version to int. err: %s", err.Error())
+		return errDetectingIptablesVersion
+	}
+
+	if majorVersion >= 5 {
+		msg := "detected iptables version on third attempt. found kernel version >= 5. NPM will use iptables-nft"
+		klog.Info(msg)
+		metrics.SendLog(util.IptmID, msg, metrics.DonotPrint)
+		util.SetIptablesToNft()
+		return nil
+	}
+
+	msg := "detected iptables version on third attempt. found kernel version < 5. NPM will use iptables-legacy"
+	klog.Info(msg)
+	metrics.SendLog(util.IptmID, msg, metrics.DonotPrint)
+	util.SetIptablesToLegacy()
+	return nil
+}
+
+// clenaupOtherIptablesChains cleans up legacy tables if using nft and vice versa.
+// It will only return an error if it fails to delete a jump rule and flush the AZURE-NPM chain (see comment about #3088 below).
+// Cleanup logic:
+// 1. delete jump rules to AZURE-NPM
+// 2. flush all chains
+// 3. delete all chains
+func (pMgr *PolicyManager) cleanupOtherIptables() error {
+	hadNFT := util.Iptables == util.IptablesNft
+	if hadNFT {
+		klog.Info("detected nft iptables. cleaning up legacy iptables")
```

Comment:
for some reason i am uncomfortable with this. But cannot think of an alternative to propose,

Reply:
Also, what about cleaning up NFT if the current is NFT, will that be taken care of in bootup ?

Reply:
> Also, what about cleaning up NFT if the current is NFT, will that be taken care of in bootup ?

Yes